### ISSUE TITLE
TransformerEncoder mask only applies to attention layers

### ISSUE THREAD

**Issue by greenpip** [pytorch/pytorch#86214]
### 🐛 Describe the bug

In the PyTorch TransformerEncoder, the mask only applies to the MultiheadAttention layers. This can result in the ith entry of the output depending on the ith entry of the input even when the i,i entry of the mask is -inf. This appears to match the description of the masking in the paper the implementation is based on. Would it make sense to offer an option to mask all data flow in the network according to the mask or to warn when the data flow in the network may not correspond to the mask, since such instances usually occur when the user is doing something wrong?

Minimal Repro:
```
from torch.autograd.functional import jacobian

# each token of output is only allowed to depend on zeroth token of input
mask = torch.zeros(5,5)
mask[:,1:] = float('-inf')

tkn_len = 50

tel = nn.TransformerEncoderLayer(tkn_len,1)
tenc = nn.TransformerEncoder(tel, num_layers=1)

print(jacobian(lambda x: tenc(x, mask=mask),
               torch.ones(5,tkn_len)
     ).abs().sum(axis=(1,3)).squeeze()
)
# The output has large nonzero entries on the diagonal.
```

### Versions

PyTorch 1.12.1, running on CPU only

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by erichan1** [comment]
Hi! 
I believe we treat masks as booleans. So -inf == True and 0 == False. We mask out a position if True, so I believe your mask might be reversed from what you intended. Can you try swapping the mask or using booleans?

I think if you mask out a row in the mask arg or if you mask out a token in src_key_padding_mask, the final output should end up having a zero in the position corresponding to the masked token. Might be wrong though - had to work through the matmuls pretty quick.

**Comment by greenpip** [comment]
According to the MultiheadAttention documentation, "For a float mask, the mask values will be added to the attention weight." So setting the i,j entry of the mask to -inf means that in the Multihead attention layer, output token i will not depend on input token j. However, there is also a residual connection that goes around the Multihead attention layer. After the residual is added, the result will have token output i depend on token input i regardless of any masking done in the MultiheadAttention layer. This is in a sense, expected behavior, since that is how the paper does it. But it might be nice to warn or something when this kind of conflict arises.

**Comment by greenpip** [comment]
change close reason

**Comment by greenpip** [comment]
changing close reason
